app/models/manageiq/providers/openstack/infra_manager/host.rb

```diff
@@ -412,11 +412,119 @@ def provide(taskid = nil)
       begin
         MiqEvent.raise_evm_job_event(self, :type => "provide", :suffix => "complete")
       rescue => err
-        $log.warn("Error raising complete provide event for #{log_target}: #{err.message}")
+        _log.warn("Error raising complete provide event for #{log_target}: #{err.message}")
       end
     end
 
     task.update_status("Finished", task_status, "Provide Complete with #{workflow_state}") if task
-    $log.info("Provide #{log_target}...Complete - Timings: #{t.inspect}")
+    _log.info("Provide #{log_target}...Complete - Timings: #{t.inspect}")
+  end
+
+  def supports_start?
+    (state.casecmp("off") == 0 && hardware.provision_state == "active")
+  end
+
+  def validate_start
+    if supports_start?
+      {:available => true,   :message => nil}
+    else
+      {:available => false,  :message => _("Cannot start. Already on.")}
+    end
+  end
+
+  def start(userid = "system")
+    ironic_set_power_state_queue(userid, "power on", "Starting", "start", :host_start)
+  end
+
+  def supports_stop?
+    state.casecmp("on") == 0
+  end
+
+  def validate_stop
+    if supports_stop?
+      {:available => true,   :message => nil}
+    else
+      {:available => false,  :message => _("Cannot stop. Already off.")}
+    end
+  end
+
+  def stop(userid = "system")
+    ironic_set_power_state_queue(userid, "power off", "Stopping", "stop", :host_stop)
+  end
+
+  def ironic_set_power_state_queue(userid = "system",
```

Comment:
This class is getting pretty long, could we place the methods in the power operations mixin? Like we do e.g. for vm?